src/webui/webapplication.cpp

```diff
@@ -173,45 +102,123 @@ namespace
+                translation.replace("'", "&#39;");
+                translation.replace("\"", "&#34;");
+
+                data.replace(i, regex.matchedLength(), translation);
+                i += translation.length();
+            }
+            else {
+                found = false; // no more translatable strings
+            }
+        }
+    }
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
+    , m_isAltUIUsed(false)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    for (int i = metaObject()->methodOffset(); i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (QLatin1String(method.tag()) == QLatin1String("WEBAPI_PUBLIC"))
+            m_publicAPIs << method.name();
+        else if (method.tag() == QLatin1String("WEBAPI_PRIVATE"))
+            m_privateAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::sendWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
+        status(500, "Internal Server Error");
+        return;
+    }
+
+    if ((!m_isAltUIUsed) && (request().path.startsWith(PATH_PREFIX_IMAGES))) {
+        sendFile(QLatin1String(":/icons/") + request().path.mid(PATH_PREFIX_IMAGES.size()));
+        header(Http::HEADER_CACHE_CONTROL, MAX_AGE_MONTH);
+        return;
+    }
+
+    const QString path {
+        (request().path != QLatin1String("/")
+                ? request().path
+                : (sessionActive()
+                   ? QLatin1String("/index.html")
+                   : QLatin1String("/login.html")))
+    };
+
+    QString localPath {
+        m_rootFolder
+                + (sessionActive() ? PRIVATE_FOLDER : PUBLIC_FOLDER)
+                + path
+    };
+
+    QFileInfo fileInfo {localPath};
+
+    if (!fileInfo.exists() && sessionActive()) {
+        // try to send public file if there is no private one
+        localPath = m_rootFolder + PUBLIC_FOLDER + path;
+        fileInfo.setFile(localPath);
+    }
+
+    while (fileInfo.filePath() != m_rootFolder) {
+        if (fileInfo.isSymLink()) {
+            status(500, "Internal Server Error");
+            print(tr("Symlinks inside WWW root folder are forbidden."), Http::CONTENT_TYPE_TXT);
+            return;
+        }
+
+        fileInfo.setFile(fileInfo.path());
+    }
+
+    sendFile(localPath);
```

Comment:
And after testing for symlinks, please check that the path does not point to a special file (socket, device, etc). As Qt does not cover that, here is a snippet for POSIX (or you can use boost::filesystem):
```c++
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#include <string.h> // for strerror()

bool isRegularFile(const char* path)
{
     struct ::stat st;
     if (!::stat(path, &st)) {
            //  analyse erno and log the error
           const auto err = errno;
           qDebug() << "Could not get file stats for path'" << path << "'. Error: " << strerror(err);
           return false;
     }
     return (st.st_mode & S_IFMT) == S_IFREG;
}
```

Reply:
@evsh, does your test function cover all cases on UNIX? I mean, should we still checking for symlinks separately?

Reply:
>please check that the path does not point to a special file (socket, device, etc). 

is it really necessary?

Reply:
> is it really necessary?

Sure, one can put there a device file for a whole disk or keyboard and read them.

Reply:
> @evsh, does your test function cover all cases on UNIX?

It is written in accordance to POSIX, thus all compatible systems (all *NIXEs) should be covered. 

> I mean, should we still checking for symlinks separately?

You do check for directory symlinks above, while this snippet tests files only. If you want to test for directories, S_IFDIR is the flag:
```c++
((st.st_mode & S_IFMT) == S_IFREG) || ((st.st_mode & S_IFMT) == S_IFDIR)
```
There are helper macros:
```c++
S_ISREG(st.st_mode) || S_ISDIR(st.st_mode)
```
that do exactly the same, but they were introduced in 1997 or something like that. So if you don't want to target older systems, they are useful (BTW, there is also  similar S_ISLNK macro that test for symbolic links).

Reply:
>Sure, one can put there a device file for a whole disk or keyboard and read them.

I'm really interested in this :)
AFAIK:
You'll need higher privilege than a normal user to create a block/char file, but there is no point in doing so, you still need to access `/dev` to read the hardware.
As for named pipes, you'll need to feed it from somewhere else; if this is the case, then the attacker can already issue shell commands (to create the pipe), no point defending it either.

There is no harm for checking more strictly, I won't insist not adding it.

Reply:
One can give many such files to you in a web UI archive. If one of them contains a proper device numbers, an intruder can use it.

Reply:
>If one of them contains a proper device numbers, an intruder can use it.

OK, but one must be insane to let these files have root permission...

Reply:
Even reading from such files would create a security problem, so root permission is not needed.

Reply:
>Even reading from such files would create a security problem, so root permission is not needed.

No, you need root to create/copy such file.

Reply:
Yes, but is it impossible to convince to convince somebody to unpack with sudo? Or install such a file by other means?

Reply:
>Yes, but is it impossible to convince to convince somebody to unpack with sudo? Or install such a file by other means?

That's what I mean by "OK" in my post above ;)